File: Problem Set 6/credit.py
import re
from cs50 import get_int

#declaring variables
num = get_int("Number: ")
len_num = len(str(num))
x = num
sum1 = 0
sum2 = 0
twoDigits = 0
sumfinal = sum1 + sum2

temp = 0

#If the card number length is incorrect
if len_num < 13 or len_num > 16:
    print("INVALID")

#Checksum algorithm
for i in range(round(len_num)):
    sum1 = sum1 + (x % 10)
    x = x / 10
    twoDigits = (2 * (x % 10))

    if twoDigits > 9:
        temp = temp + (twoDigits % 10)
        twoDigits /= 10
        temp = temp + twoDigits
        sum2 = sum2 + temp
        temp = 0

    else:
        sum2 = sum2 + twoDigits
        x /= 10

#If checksum is not correct, the number is "INVALID"
if sumfinal % 10 != 0:
    print("INVALID")

#Check for AMEX
if re.search("^34", str(num)) or re.search("^37", str(num)):
    if len_num == 15:
        print("AMEX")

#Check for MasterCard
if re.search("^51", str(num)) or re.search("^52", str(num)) or re.search("^53", str(num)) or re.search("^54", str(num)) or re.search("^55", str(num)):
    if len_num == 16:
        print("MASTERCARD")
        
#Check for VISA
if re.search("^4", str(num)):
    if len_num == 13 or len_num == 16:
        print("VISA")

File: Problem Set 1/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

int main(void)
{
    //input function
    long cardNumber = get_long("What's your credit card number: ");
    long x = cardNumber;
    int length = 0;
    int sum1 = 0;
    int sum2 = 0;
    int temp = 0;
    long power = cardNumber;
    long visa = 0;
    int check = 0; //for card Numbers with last digit zero, but dont qualify other checks

    //finding length of credict card number algorithm
    for (long count = 1; count <= cardNumber; count *= 10)
        {
        length++;
        }

    if(length < 13 || length > 16)
        {
        printf("EOF\n");
        }

    //checksum algorithm
    for (int i = 0; i <= (length / 2) + 1; i++)
        {
        sum1 = sum1 + (x % 10);
        x = x / 10;
        int twoDigits = (2 * (x % 10));
        
        if (twoDigits > 9)
            {
            temp = temp + (twoDigits % 10);
            twoDigits /= 10;
            temp = temp + twoDigits;
            sum2 = sum2 + temp;
            temp = 0;
            }
        
        else
            {
            sum2 = sum2 + twoDigits;
            }

        x = x / 10;   
    }
        
    int sumfinal = sum1 + sum2;

    if (sumfinal % 10 == 0)
        {
        visa = power / (pow (10, (length - 1)));
        power = power / (pow (10, (length - 2)));
        
        //American Express
        if (length == 15 && (power == 34 || power == 37)) 
            {
            printf("AMEX\n");
            check ++;
            }

        //MasterCard
        if (length == 16 && (power == 51 || power == 52 || power == 53 || power == 54 || power == 55))
            {
            printf("MASTERCARD\n");
            check++;
            }

        //Visa
        if  (visa == 4 && (length == 13 || length == 16))
            {
            printf("VISA\n");
            check++;
            }        
        }
        
        else
            {
            printf("INVALID\n");
            }

        if  (check == 0)
            {
            printf("INVALID\n");
            }
}
